### PR TITLE
Add streaming support for Advanced rate limiter & handle SSE Error events

### PR DESCRIPTION
**Pull request by Thushani-Jayasekera**
$subject

- https://github.com/wso2/api-platform/issues/1563
- https://github.com/wso2/api-platform/issues/1548
- https://github.com/wso2/api-platform/issues/1386
- 
## Overview

The SDK (`v0.2.0`) already defines `StreamingResponsePolicy` with `OnResponseBodyChunk` /
`NeedsMoreResponseData` and `BodyModeStream`. The work is entirely in the policy itself.

**Core principle:** The operator does not choose between streaming and non-streaming. The
`response_body` source type handles both transparently. LLM providers always use
`Transfer-Encoding: chunked`, so `OnResponseBodyChunk` fires regardless of whether the
client requested SSE or a plain JSON response. The policy detects which format it is on the
first chunk and adapts automatically.

---

## 1. No New Source Type

The existing `response_body` source type is used for both cases. No `response_body_stream`
type is needed. Configuration stays the same as today:

```yaml
costExtraction:
  enabled: true
  sources:
    - type: response_body
      jsonPath: usage.total_tokens  # same path for both SSE and plain JSON responses
      multiplier: 0.000002          # applied to the extracted numeric value
  default: 1                        # fallback cost when extraction finds nothing
```

---

## 2. Updated `Mode()` — Stream Instead of Buffer for `response_body`

```go
// ratelimit.go

func (p *RateLimitPolicy) Mode() policy.ProcessingMode {
    requestBodyMode  := policy.BodyModeSkip
    responseBodyMode := policy.BodyModeSkip

    if p.requiresRequestBody() {
        requestBodyMode = policy.BodyModeBuffer
    }

    if p.hasResponseBodySource() {
        // Switch to BodyModeStream for response_body cost sources.
        // The kernel enters FULL_DUPLEX_STREAMED mode only when every policy in the
        // chain agrees on BodyModeStream. If any other policy requests BodyModeBuffer,
        // the kernel falls back to buffered mode and calls OnResponseBody instead —
        // which already handles both SSE and plain JSON via extractFromBodyBytes.
        responseBodyMode = policy.BodyModeStream
    }

    return policy.ProcessingMode{
        RequestHeaderMode:  policy.HeaderModeProcess,
        RequestBodyMode:    requestBodyMode,
        ResponseHeaderMode: policy.HeaderModeProcess,
        ResponseBodyMode:   responseBodyMode,
    }
}

// hasResponseBodySource returns true if any quota has a response_body cost source.
// This is used by Mode() to decide whether to request BodyModeStream for the response.
func (p *RateLimitPolicy) hasResponseBodySource() bool {
    for _, q := range p.quotas {
        if q.CostExtractionEnabled && q.CostExtractor != nil {
            for _, src := range q.CostExtractor.GetConfig().Sources {
                if src.Type == CostSourceResponseBody {
                    return true
                }
            }
        }
    }
    return false
}
```

**Kernel behavior:**
- If every policy in the chain returns `BodyModeStream` (or `BodyModeSkip`), the kernel enters
  `FULL_DUPLEX_STREAMED` mode and calls `OnResponseBodyChunk` per chunk.
- If any other policy requires `BodyModeBuffer`, the kernel falls back to buffered mode and
  calls `OnResponseBody` instead. The existing `extractFromBodyBytes` / `extractFromSSEBodyBytes`
  fallback already handles both SSE and JSON in that path — no changes needed there.

---

## 3. Per-Request Streaming State

Since `RateLimitPolicy` is a singleton shared across all requests, per-stream state must live
in `SharedContext.Metadata` (which is per-request and persists across all chunk calls for the
same request).

```go
// ratelimit.go

const rateLimitStreamStateKey = "ratelimit:stream_state"

// streamQuotaState holds per-stream accumulation state for one quota.
// A separate instance is kept per quota name inside the metadata map so that
// multiple quotas on the same request track their state independently.
type streamQuotaState struct {
    lastCost    float64 // most recently extracted cost value (updated on each matching chunk)
    found       bool    // true once at least one chunk yielded a value for the jsonPath
    isSSE       *bool   // nil = format not yet determined; true = SSE, false = plain JSON
    partialLine []byte  // SSE mode only: bytes from the current chunk with no trailing newline,
                        // carried forward and prepended to the next chunk
    accumulated []byte  // JSON mode only: full body bytes gathered across all chunks until EOS
}
```

---

## 4. `OnResponseBodyChunk` — Auto-Detect SSE vs JSON

```go
// OnResponseBodyChunk is called by the kernel for every response body chunk when the chain
// runs in FULL_DUPLEX_STREAMED mode. It handles both SSE and plain JSON responses:
//   - SSE  → parse each chunk line-by-line; extract jsonPath per event; last match wins at EOS.
//   - JSON → accumulate raw bytes across chunks; extract jsonPath once at EOS.
// Format detection happens on the first non-empty chunk and is fixed for the lifetime of
// the stream. The chunk itself is always passed through unmodified.
func (p *RateLimitPolicy) OnResponseBodyChunk(
    ctx context.Context,
    respCtx *policy.ResponseStreamContext,
    chunk *policy.StreamBody,
    _ map[string]interface{},
) policy.ResponseChunkAction {

    // Load (or lazily create) the per-request state map from shared metadata.
    state := p.getOrInitStreamState(respCtx.Metadata)

    for i := range p.quotas {
        q := &p.quotas[i]
        if !q.CostExtractionEnabled || q.CostExtractor == nil {
            continue
        }

        quotaName := quotaNameFor(q, i)
        qs := state[quotaName]

        for _, src := range q.CostExtractor.GetConfig().Sources {
            if src.Type != CostSourceResponseBody {
                continue
            }

            // Detect SSE vs JSON on the first non-empty chunk.
            // SSE streams always begin with "data:" or "event:" on the first content line.
            // This detection is done once and stays fixed for the rest of the stream.
            if qs.isSSE == nil && len(chunk.Chunk) > 0 {
                trimmed := strings.TrimLeft(string(chunk.Chunk), " \r\n")
                isSSE := strings.HasPrefix(trimmed, "data:") || strings.HasPrefix(trimmed, "event:")
                qs.isSSE = &isSSE
                slog.Debug("Detected response body format",
                    "quota", quotaName, "isSSE", isSSE)
            }

            if qs.isSSE != nil && *qs.isSSE {
                // SSE mode: parse whatever complete lines exist in this chunk.
                // Any bytes after the last newline are incomplete and held in partialLine
                // to be prepended to the next chunk — guarding against a JSON payload
                // being split across a chunk boundary.
                combined := append(qs.partialLine, chunk.Chunk...)
                lastCost, found, remaining := parseSSEChunk(combined, src.JSONPath)
                qs.partialLine = remaining
                if found {
                    // Overwrite on each match: last-match-wins gives the final/EOS value,
                    // which is what LLM providers populate with the complete token count.
                    qs.lastCost = lastCost * src.Multiplier
                    qs.found = true
                }
            } else {
                // JSON mode: the full body arrives as multiple chunks due to chunked
                // transfer encoding. Accumulate all bytes and defer extraction to EOS.
                qs.accumulated = append(qs.accumulated, chunk.Chunk...)
            }
        }

        state[quotaName] = qs
    }
    // Write state back so it is visible to the next chunk call for this request.
    respCtx.Metadata[rateLimitStreamStateKey] = state

    // EOS: all chunks received. Finalize JSON extraction (SSE values are already
    // up-to-date from per-chunk parsing) and consume tokens against each quota.
    if chunk.EndOfStream {
        p.finalizeAndConsumeStreamingCosts(ctx, respCtx, state)
    }

    return policy.ResponseChunkAction{} // passthrough — chunk bytes unmodified
}

// NeedsMoreResponseData always returns false because we process each chunk as it arrives
// rather than waiting to accumulate a minimum number of bytes first.
// For SSE: we extract per-event inline.
// For JSON: we accumulate manually in streamQuotaState.accumulated and only extract at EOS.
func (p *RateLimitPolicy) NeedsMoreResponseData(_ []byte) bool {
    return false
}
```

---

## 5. EOS Finalization and Cost Consumption

```go
// finalizeAndConsumeStreamingCosts is called once at EOS (chunk.EndOfStream == true).
// For SSE quotas the cost is already in streamQuotaState from per-chunk parsing.
// For JSON quotas the accumulated bytes are extracted here for the first time.
// Tokens are consumed via ConsumeN (preferred) or ConsumeOrClampN (fallback) — the
// same pattern used in OnResponseBody for the buffered path.
func (p *RateLimitPolicy) finalizeAndConsumeStreamingCosts(
    ctx context.Context,
    respCtx *policy.ResponseStreamContext,
    state map[string]*streamQuotaState,
) {
    // quotaKeys were stored in metadata during the request phase (OnRequestHeaders /
    // OnRequestBody) so we can resolve the rate-limit key for each quota here.
    quotaKeys, _ := respCtx.Metadata[rateLimitKeysKey].(map[string]string)

    for i := range p.quotas {
        q := &p.quotas[i]
        if !q.CostExtractionEnabled || q.CostExtractor == nil {
            continue
        }
        // Only process quotas that use response_body sources; other source types
        // (response_header, response_metadata, response_cel) are handled elsewhere.
        if !p.hasResponseBodySourceForQuota(q) {
            continue
        }

        quotaName := quotaNameFor(q, i)
        key := quotaKeys[quotaName]
        qs := state[quotaName]

        var actualCost float64

        if qs != nil {
            if qs.isSSE == nil || !*qs.isSSE {
                // JSON mode: all bytes are in qs.accumulated — extract now.
                // Re-uses extractFromBodyBytes which handles standard JSONPath evaluation.
                for _, src := range q.CostExtractor.GetConfig().Sources {
                    if src.Type != CostSourceResponseBody {
                        continue
                    }
                    if cost, ok := extractFromBodyBytes(qs.accumulated, src.JSONPath); ok {
                        qs.lastCost = cost * src.Multiplier
                        qs.found = true
                    }
                }
            }
            // SSE mode: qs.lastCost was populated incrementally during OnResponseBodyChunk.
            if qs.found {
                actualCost = qs.lastCost
            } else {
                // No chunk matched the jsonPath — fall back to the configured default.
                actualCost = q.CostExtractor.GetConfig().Default
            }
        } else {
            // State was never initialised (e.g. all chunks were empty). Use default.
            actualCost = q.CostExtractor.GetConfig().Default
        }

        if actualCost < 0 {
            actualCost = 0
        }
        if actualCost == 0 {
            // Zero cost: nothing to consume. Skip to avoid unnecessary limiter calls.
            continue
        }

        // Consume tokens. ConsumeN (CostTracker) records accurate overflow/consumed
        // metrics. ConsumeOrClampN is the safe fallback for limiters that don't implement
        // CostTracker — it clamps the deduction to the available balance.
        if tracker, ok := q.Limiter.(limiter.CostTracker); ok {
            tracker.ConsumeN(ctx, key, int64(actualCost))
        } else {
            q.Limiter.ConsumeOrClampN(ctx, key, int64(actualCost))
        }

        slog.Debug("Streaming EOS cost consumed",
            "quota", quotaName, "key", key, "cost", actualCost,
            "isSSE", qs != nil && qs.isSSE != nil && *qs.isSSE)
    }
}
```

---

## 6. SSE Chunk Parser (Partial-Line Safe)

Added to `cost_extractor.go`. Reuses the existing `sseDataPrefix`, `sseEventPrefix`, `sseDone`
constants.

```go
// parseSSEChunk scans buf for complete SSE lines (delimited by '\n') and applies
// jsonPath to the JSON payload of each data:/event: line.
//
// It returns:
//   - lastCost: the numeric value from the last line that matched jsonPath
//   - found:    true if at least one line matched
//   - remaining: any bytes after the last '\n' (an incomplete line); the caller must
//                prepend these to the next chunk before calling parseSSEChunk again
//
// last-match-wins is intentional: LLM providers send the authoritative token count in
// the final usage event, so overwriting on each match naturally yields the right value.
func parseSSEChunk(buf []byte, jsonPath string) (lastCost float64, found bool, remaining []byte) {
    s := string(buf)
    for {
        idx := strings.IndexByte(s, '\n')
        if idx < 0 {
            // No complete line remaining — return the fragment so the caller can
            // prepend it to the next chunk.
            return lastCost, found, []byte(s)
        }
        line := strings.TrimRight(s[:idx], "\r") // strip optional \r from CRLF
        s = s[idx+1:]

        var value string
        if strings.HasPrefix(line, sseDataPrefix) {
            value = strings.TrimPrefix(line, sseDataPrefix)
        } else if strings.HasPrefix(line, sseEventPrefix) {
            value = strings.TrimSpace(strings.TrimPrefix(line, sseEventPrefix))
        } else {
            continue // comment, id:, retry:, or blank separator line — ignore
        }

        if value == sseDone || value == "" {
            continue // stream terminator or empty data field — skip
        }

        valueStr, err := utils.ExtractStringValueFromJsonpath([]byte(value), jsonPath)
        if err != nil {
            // jsonPath not present in this event (e.g. content chunk with no usage field).
            // This is expected for most SSE events — only trailing events carry usage.
            continue
        }
        cost, err := strconv.ParseFloat(valueStr, 64)
        if err != nil {
            continue
        }
        lastCost = cost
        found = true // will be overwritten if a later line also matches
    }
}
```

---

## 7. `OnResponseBody` — Buffered Fallback (No Changes)

When the kernel falls back to buffered mode (another policy in the chain requires
`BodyModeBuffer`), `OnResponseBody` is called with the complete buffered body.
`extractFromBodyBytes` already tries plain JSON first, then `extractFromSSEBodyBytes`
as a fallback — no changes required.

---

## 8. Rate Limit Headers in Streaming Mode

Response headers are already committed to the client before the first chunk arrives.
`X-RateLimit-*` headers set inside `OnResponseBodyChunk` are ignored by Envoy. This is
a hard constraint of the streaming protocol.

**Mitigations:**
- The `OnRequestHeaders` pre-check still blocks requests when the quota is already exhausted.
- Rate limit headers from the request phase reflect the pre-request quota state, which is
  the closest approximation possible.
- Post-stream token state is reflected accurately on the *next* request.

---

## 9. Metadata & CEL

At EOS in `finalizeAndConsumeStreamingCosts`, write the extracted value into
`SharedContext.Metadata` (e.g. `ratelimit:stream_usage:{quotaName}`) so downstream
analytics policies can read it.

`response_cel` sources are unaffected — they continue using buffered mode (`requiresResponseBody()`
remains true for `response_cel`) and evaluate against the full body in `OnResponseBody`.

---

## 10. Provider Compatibility

| Provider | Format | `jsonPath` |
|---|---|---|
| **OpenAI** | SSE; last `data:` before `data: [DONE]` has `usage.total_tokens` | `usage.total_tokens` |
| **Anthropic** | SSE; `event: message_delta` has `usage.output_tokens` | `usage.output_tokens` |
| **AWS Bedrock** | SSE; final chunk has `amazon-bedrock-invocationMetrics.totalTokenCount` | `amazon-bedrock-invocationMetrics.totalTokenCount` |
| **Any non-streaming LLM** | Chunked JSON; accumulated and extracted at EOS | same `jsonPath` as before |

The last-match-wins strategy works for all SSE providers: usage appears only in trailing events.
For plain JSON sent as chunked transfer, bytes are accumulated and extracted at EOS identically
to the current buffered path.

---

## 11. Edge Case Handling

| Case | Handling |
|---|---|
| Usage field absent in all chunks | `found == false` → use `costExtraction.default` |
| Partial JSON spanning two chunks | `parseSSEChunk` holds remainder in `partialLine`; prepended to next chunk |
| Multiple `data:` lines with usage | Last match wins — correct for all known providers |
| Keep-alive / blank lines | Skipped (`value == ""` check) |
| `data: [DONE]` terminator | Skipped (`value == sseDone` check) |
| First chunk is empty (keep-alive) | Detection deferred until first non-empty chunk |
| Mixed SSE + JSON in same response | Not expected; detection on first non-empty chunk is authoritative for the stream |
| `response_cel` source alongside `response_body` | `response_cel` forces `BodyModeBuffer`; kernel calls `OnResponseBody`; `extractFromBodyBytes` SSE fallback handles both |

---

## 12. Files to Change

| File | Change |
|---|---|
| `ratelimit.go` | `hasResponseBodySource()`, updated `Mode()`, `streamQuotaState`, `getOrInitStreamState`, `OnResponseBodyChunk`, `NeedsMoreResponseData`, `finalizeAndConsumeStreamingCosts` |
| `cost_extractor.go` | Add `parseSSEChunk` (reuses existing SSE constants) |
| `cost_extractor_test.go` | Tests for `parseSSEChunk`: partial lines, multi-chunk, missing field, plain JSON accumulation |
| `ratelimit_integration_test.go` | End-to-end: mock SSE stream → verify EOS consumption; mock chunked JSON → verify accumulation |
| `policy-definition.yaml` | No schema change — `response_body` already exists |

## Summary by CodeRabbit

* **New Features**
  * Streaming response support with incremental SSE/JSON parsing for per-chunk cost extraction and interim availability snapshots
  * Delegation hooks so provider-specific handlers can process response chunks
  * Upstream rate-limit headers are removed/sanitized
  * Guardrails now pass through upstream guardrail errors, append stream terminators on intervention, and suppress subsequent chunks after a violation

* **Tests**
  * Added unit and integration tests covering multi-chunk SSE, chunked JSON accumulation, edge cases, multiplier handling, and delegate behavior

### CODE REVIEW

**Review by renuka-fernando — DISMISSED** [review]

File: policies/advanced-ratelimit/cost_extractor.go [hunk 1]
```diff
@@ -482,6 +482,64 @@ func extractFromSSEBodyBytes(bodyBytes []byte, jsonPath string) (float64, bool)
...
+//   - lastCost:  the numeric value from the last line that matched jsonPath
+//   - found:     true if at least one line yielded a value
+//   - remaining: any bytes after the last newline (an incomplete line); the caller
+//                must prepend these to the next chunk before calling parseSSEChunk
+//                again to guard against JSON payloads split across chunk boundaries
+//
+// last-match-wins is intentional: LLM providers send the authoritative token count
+// in the final usage event, so overwriting on each match naturally yields the right
+// value without needing to know which event is "last" in advance.
+func parseSSEChunk(buf []byte, jsonPath string) (lastCost float64, found bool, remaining []byte) {
```
Comment (renuka-fernando) [inline policies/advanced-ratelimit/cost_extractor.go:498]: This is specific to Token Based Rate Limit noh, shouldn't it be come to Token-Based RL. As this is a private func this would be fine for now.

**Review by renuka-fernando — DISMISSED** [review]

File: policies/url-guardrail/urlguardrail_test.go [hunk 2]
```diff
@@ -481,3 +481,130 @@ func TestOnRequestBodyAndOnResponseBody(t *testing.T) {
...
+
+func newStreamingURLPolicy(timeout int) *URLGuardrailPolicy {
+	return &URLGuardrailPolicy{
+		hasResponseParams: true,
+		responseParams: URLGuardrailPolicyParams{
+			Enabled:           true,
+			StreamingJsonPath: DefaultStreamingJsonPath,
+			Timeout:           timeout,
+		},
+	}
```
[the remark this answers is not in the document]
Comment (Thushani-Jayasekera) [reply policies/url-guardrail/urlguardrail_test.go:526]: The described code doesn't exist in the current files. Searching for URLGuardrailPolicyParams, StreamingJsonPath, DefaultResponseJsonPath,
   and JsonPath returns no matches in either urlguardrail.go or urlguardrail_test.go.                                                       
                                                                                                                                            
  The newStreamingURLPolicy function at line 518 uses a struct literal with fields Enabled, StreamingJsonPath, and Timeout — but those      
  symbols don't resolve in the current source (the grep found no matches for them in the production file either). This suggests the test    
  file references types/constants that don't yet exist in urlguardrail.go.                                                                  
                                                                                                                                            
  Nothing to fix — the prerequisite types (URLGuardrailPolicyParams, DefaultStreamingJsonPath, DefaultResponseJsonPath) are not present in  
  the codebase. Do you want me to implement these, or is there a branch/PR where this struct is being added?

**Review by renuka-fernando — APPROVED** [review]
